src/python/pants/core/goals/repl.py

```diff
@@ -139,8 +147,10 @@ async def run_repl(
             InteractiveProcess(
                 argv=request.args,
                 env=env,
-                run_in_workspace=True,
+                input_digest=request.digest,
```

Comment:
@Eric-Arellano  / @stuhood: This line uncovered the issue with the Python repls. The `repl` goal was not passing the input root through.

Reply:
Mmm.

The repl should probably be doing whatever we do for `run` [here](https://github.com/pantsbuild/pants/blob/66673f027129fe04c6e5c336429fa793108cc0e2/src/python/pants/core/goals/run.py#L139-L160)...? Or we should add explicit support to `InteractiveProcess` for materializing in a temporary subdir? But it's not clear that that removes very much boilerplate.

Disabling "in workspace" for the `repl` would be confusing I think... that's probably a blocker.

Reply:
> Or we should add explicit support to InteractiveProcess for materializing in a temporary subdir? 

`InteractiveProcess` already has the ability to materialize in a temporary directory. That is what its `input_digest` attribute enables. But it requires `run_in_workspace=False`. The Python repls were providing an input root, but the core repl goal was ignoring it when constructing the `InteractiveProcess` and hard-coding `run_in_workspace=True`.

Reply:
Why does `run` create a separate temporary subdirectory?

Reply:
> but the core repl goal was ignoring it when constructing the InteractiveProcess and also hard-coding run_in_workspace=True.

They're not. If you look right above, they are writing the digest into the Workspace's temp dir. 

We use the build root for both `run` and `repl` because we want to give users the flexibility of accessing the build root. For example, `run` needs to be able to write files to the build root, without forcing users to have their scripts use absolute paths to the build root. For `repl`, our argument was that it's for experimentation so you should be able to do things like `open("f.ext")` even if you don't have a `file`/`files` target for that.

If Scala cannot run in the build root because append only caches, such it is. Your new `ReplRequest.run_in_workspace` field is a good one though, and I think here in `core/goals/repl.py` you will want to use an `if` statement: if running in workspace, use the temp dir stuff, unmodified from before. Else, use `input_digest` and `append_only_caches` via the engine's mechanism.

Reply:
Simpler solution was just to set `input_digest` to EMPTY_DIGEST (similar to the existing code which doesn't pass it) if `run_in_workspace=True` since the data in the input_digest would have already been written to the temporary directory. The temporary directory still exists even if `run_in_workspace=False` since it is passed to the rules generating the `ReplRequest` and I didn't want to disturb that API contract.